sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java

```diff
@@ -0,0 +1,302 @@
+      bulkRequestBuilder.add(client.prepareIndex(ES_INDEX, ES_TYPE, null).setSource(source));
+    }
+    final BulkResponse bulkResponse = bulkRequestBuilder.execute().actionGet();
+    if (bulkResponse.hasFailures()) {
+      throw new IOException("Cannot insert samples in index " + ES_INDEX);
+    }
+  }
+
+  @Test
+  public void testSizes() throws Exception {
+    sampleIndex(NB_DOCS);
+    PipelineOptions options = PipelineOptionsFactory.create();
+    String[] addresses = {"http://" + ES_IP + ":" + esHttpPort};
+    ElasticsearchIO.Read read =
+        ElasticsearchIO.read().withConnectionConfiguration(
+            ElasticsearchIO.ConnectionConfiguration
+                .create(addresses, ES_INDEX, ES_TYPE));
+    BoundedElasticsearchSource initialSource =
+        new BoundedElasticsearchSource(read, null);
+    // can't use equal assert as Elasticsearch indexes never have same size
+    // (due to internal Elasticsearch implementation)
+    long estimatedSize = initialSource.getEstimatedSizeBytes(options);
+    LOGGER.info("Estimated size: {}", estimatedSize);
+    assertTrue("Wrong estimated size", estimatedSize > 40000);
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testRead() throws Exception {
+    sampleIndex(NB_DOCS);
+
+    TestPipeline pipeline = TestPipeline.create();
+
+    String[] addresses = { "http://" + ES_IP + ":" + esHttpPort };
+    PCollection<String> output = pipeline.apply(
+        ElasticsearchIO.read().withConnectionConfiguration(
+            ElasticsearchIO.ConnectionConfiguration
+                .create(addresses, ES_INDEX, ES_TYPE))
+            .withScrollKeepalive("5m"));
+    PAssert.thatSingleton(output.apply("Count", Count.<String>globally())).isEqualTo(NB_DOCS);
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testReadWithQuery() throws Exception {
+    sampleIndex(NB_DOCS);
+
+    String query = "{\n"
+        + "  \"query\": {\n"
+        + "  \"match\" : {\n"
+        + "    \"scientist\" : {\n"
+        + "      \"query\" : \"Einstein\",\n"
+        + "      \"type\" : \"boolean\"\n"
+        + "    }\n"
+        + "  }\n"
+        + "  }\n"
+        + "}";
+
+    Pipeline pipeline = TestPipeline.create();
+
+    String[] addresses = {"http://" + ES_IP + ":" + esHttpPort};
+    PCollection<String> output = pipeline.apply(
+        ElasticsearchIO.read().withConnectionConfiguration(
+            ElasticsearchIO.ConnectionConfiguration
+                .create(addresses, ES_INDEX, ES_TYPE))
+            .withQuery(query));
+    PAssert.thatSingleton(output.apply("Count", Count.<String>globally())).isEqualTo(NB_DOCS / 10);
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWrite() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+
+    String[] addresses = {"http://" + ES_IP + ":" + esHttpPort};
+    String[] scientists =
```

Comment:
Extract this (creating the `data` list) into a helper method to dedup it against `sampleIndex`?
E.g. create a method `List<String> createDocuments(int numDocs)` and a separate method `createIndexWithDocuments(List<String> docs)`